frame/staking/src/offchain_election.rs

```diff
@@ -155,19 +163,165 @@ pub fn get_balancing_iters<T: Trait>() -> usize {
 	}
 }
 
+/// Find the maximum `len` that a compact can have in order to fit into the block weight.
+///
+/// This only returns a value between zero and `size.nominators`.
+pub fn maximum_compact_len<W: crate::WeightInfo>(
+	winners_len: u32,
+	size: ElectionSize,
+	max_weight: Weight,
+) -> u32 {
+	use sp_std::cmp::Ordering;
+	if size.nominators < 1 {
+		return size.nominators;
+	}
+
+	let weight_with = |voters: u32| -> Weight {
+		W::submit_solution_better(
+			size.validators.into(),
+			size.nominators.into(),
+			voters,
+			winners_len,
+		)
+	};
+
+	let mut voters = size.nominators.max(2);
+	let mut step = voters;
+	while voters <= size.nominators.max(2) {
```

Comment:
This loop still seems a bit overcomplicated for me and there are some weird cases with `break` statements.
Imagine we get to `step==0` point, then reach the first `break`. We will end up executing `current.cmp(max_weight)` anyway, even though `current` might be calculated with a different `voters` value and then end up altering `voters` again.

I took a stab at simplifying it a bit and cleaning up the corner cases, wdyt?
```rust
let max_voters = size.nominators.max(2);
let mut voters = max_voters;

let next_voters = |current_weight: Weight, voters: u32, step: u32| -> Result<u32, ()> {
    match current.cmp(&max_weight) {
        Ordering::Less => {
            let next_voters = voters.checked_add(step);
            match next_voters {
                Some(voters) if voters < max_voters => Ok(voters),
                _ => Err(()),
            }
        },
        Ordering::Greater => voters.checked_sub(step).ok_or(()),
        Ordering::Equal => Ok(voters),
    }
}

// First binary-search the right amount of voters
let mut step = voters / 2;
while step > 0 {
    let current_weight = weight_with(voters);
    match next_voters(current_weight, voters, step) {
        // proceed with the binary search
        Ok(next) if next != voters => {
            voters = next;
        },
        // we are out of bounds, break out of the loop.
        Err(()) => {
            break;
        },
        // we found the right value - early exit the function
        Ok(voters) => return voters
    }
    step = step / 2;
}

// Time to finish.
// We might have reduced less than expected due to rounding error. Reduce linearly one last time.
let mut current = weight_with(voters);
match next_voters(current, voters, 1) {
    Ok(next) => next,
    // out of bounds, just return what we have
    Err(()) => voters
}
```

The last part can also be run in a loop until `next != voters`. Basically something like this:
```rust
loop {
  let mut current = weight_with(voters);
  let next = next_voters(current, voters, 1).unwrap_or(voters);
  if next == voters { break }
  else { voters = next }
} 
```

Reply:
Yes I must admit this is much cleaner, and whispering more experienced author 😄 but I still had to add a minor complication to the last phase to make all my tests happy, and prevent infinite loops.